### PR TITLE
Issue 122 refinement types - numeric and collection predicates

### PR DESCRIPTION
**Pull request by leigh-perry**
https://github.com/zio/zio-config/issues/122

### CODE REVIEW

File: zio-config-refined/src/main/scala/zio/config/refined/internal/TypeInference.scala [hunk 1]
```diff
@@ -0,0 +1,187 @@
+package zio.config.refined
```
Comment (afsalthaj) [inline zio-config-refined/src/main/scala/zio/config/refined/internal/TypeInference.scala]: Let's change the name of TypeInference to `TypeInferenceSupport.scala`
We should also make this private as well.

File: examples/src/main/scala/zio/config/examples/RefinedReadConfig.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+object RefinedReadConfig extends App {
+  override def run(args: List[String]): ZIO[ZEnv, Nothing, Int] =
+    ZIO.accessM { env =>
+      Config
+        .fromMap(Map("LDAP" -> "ldap", "PORT" -> "1999", "DB_URL" -> "ddd"), RefinedProd.prodConfig)
+        .flatMap(config => RefinedProd.myAppLogic.provide(config))
+        .foldM(
+          failure => env.console.putStrLn(failure.toString) *> ZIO.succeed(1),
+          _ => env.console.putStrLn("👍") *> ZIO.succeed(0)
```
Comment (afsalthaj) [inline examples/src/main/scala/zio/config/examples/RefinedReadConfig.scala]: Is that `:+1` , and GitHub showing 👍

File: zio-config-refined/src/main/scala/zio/config/refined/internal/TypeInference.scala (hunk 1 above)
Comment (afsalthaj) [reply zio-config-refined/src/main/scala/zio/config/refined/internal/TypeInference.scala]: Or even PariallyAppliedTypes.scala

The word `TypeInference` is too generic.

**Comment by afsalthaj** [comment]
Once you make the slight changes, it all looks good to me 👍

File: examples/src/main/scala/zio/config/examples/RefinedReadConfig.scala (hunk 2 above)
Comment (leigh-perry) [reply examples/src/main/scala/zio/config/examples/RefinedReadConfig.scala]: No, it is the real emoji

File: zio-config-refined/src/main/scala/zio/config/refined/internal/TypeInference.scala (hunk 1 above)
Comment (leigh-perry) [reply zio-config-refined/src/main/scala/zio/config/refined/internal/TypeInference.scala]: Agreed

File: examples/src/main/scala/zio/config/examples/RefinedReadConfig.scala (hunk 2 above)
Comment (afsalthaj) [reply examples/src/main/scala/zio/config/examples/RefinedReadConfig.scala]: 👍

**Pull request merged**